### PR TITLE
Make build info tolerant to json errors

### LINKED ISSUE #49754: Failure to parse tsconfig.tsbuildinfo throwss hard-to-find exception
# Bug Report
If the `tsconfig.tsbuildinfo` file is corrupt (in my case due to a system power fail while compiling), a JSON parsing exception is thrown. It's not clear where the error comes from and it took some time to find out it was the `tsconfig.tsbuildinfo`.

Please improve this by including information on what piece of JSON was corrupt when a parsing error occurs.
### 🔎 Search Terms

`tsconfig` `json` `parse` `corrupt`

### 🕗 Version & Regression Information

Version 4.7.4
- This is a crash

### 🙁 Actual behavior
A "SyntaxError: Unexpected token x in JSON at position 0" message that is not very informative.

### 🙂 Expected behavior
An indication what was parsed along with the parsing error message

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #49754

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -5360,20 +5360,16 @@ namespace ts {
         return getStringFromExpandedCharCodes(expandedCharCodes);
     }
 
+    export function readJsonOrUndefined(path: string, hostOrText: { readFile(fileName: string): string | undefined } | string): object | undefined {
+        const jsonText = isString(hostOrText) ? hostOrText : hostOrText.readFile(path);
+        if (!jsonText) return undefined;
```
Comment (amcasey) [inline src/compiler/utilities.ts:5365]: Is this going to return `undefined` when `jsonText === ""` and, if so, is that desirable?

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -5398,6 +5398,10 @@
         "category": "Message",
         "code": 6400
     },
+    "Project '{0}' is out of date because there was error reading file '{1}'": {
+        "category": "Message",
+        "code": 6411
```
Comment (amcasey) [inline src/compiler/diagnosticMessages.json:5403]: Was this supposed to be 6401?

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -805,7 +804,7 @@ namespace ts {
         const emitHost: EmitHost = {
             getPrependNodes: memoize(() => [...prependNodes, ownPrependInput]),
             getCanonicalFileName: host.getCanonicalFileName,
-            getCommonSourceDirectory: () => getNormalizedAbsolutePath(buildInfo.bundle!.commonSourceDirectory, buildInfoDirectory),
+            getCommonSourceDirectory: () => getNormalizedAbsolutePath(buildInfo!.bundle!.commonSourceDirectory, buildInfoDirectory),
```
Comment (amcasey) [inline src/compiler/emitter.ts:807]: Is the `!` required?  It looks like we checked for undefined above.

Comment (amcasey) [reply src/compiler/emitter.ts:807]: (And below)

File: src/compiler/tsbuild.ts [hunk 4]
```diff
@@ -16,6 +16,7 @@ namespace ts {
          */
         OutOfDateWithPrepend,
         OutputMissing,
+        ErrorReadingFile,
```
Comment (amcasey) [inline src/compiler/tsbuild.ts:19]: Nit: I might call this "ErrorReadingBuildInfo"

**Review by amcasey — APPROVED** [review]

File: src/compiler/emitter.ts (hunk 3 above)
Comment (sheetalkamat) [reply src/compiler/emitter.ts:807]: arrows dont retain the narrowed values..

File: src/compiler/tsbuild.ts (hunk 4 above)
Comment (sheetalkamat) [reply src/compiler/tsbuild.ts:19]: Made this generic for future

File: src/compiler/diagnosticMessages.json (hunk 2 above)
Comment (sheetalkamat) [reply src/compiler/diagnosticMessages.json:5403]: Yes

File: src/compiler/utilities.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/compiler/utilities.ts:5365]: We expect object literal os "" is going to error later anyways so this is desirable

**Pull request merged**